Loot with chance based enchantment(s)

MC-MrBirdy opened the pull request:
Pull request as suggested by Archy-X for the Feature request "Loot with chance based enchantment(s) #380"

**In short:**
This allows a chance setting to be defined per loot enchantment, giving the option to make certain enchantments more rare.

Archy-X reviewed (changes requested):
Instead of loading the chance from the config through the split enchantment string, I believe we should support a map-based enchantment entry (in addition to the simple string) with key-values for the enchantment name, level, and chance. This way if we add more options to an enchant entry in the future, each option is properly named rather than based on an arbitrary order.

For example, we should load chance only if the enchant is in this format:
```yaml
enchantments:
- name: sharpness
  level: 3
  chance: 0.5
- 'unbreaking 3' # The current string list entry should still be supported
```

You can implement this by looping through `ConfigurationNode#childrenList` and checking `ConfigurationNode#isMap`. If it's a map, load the child keys name, level, and chance. If not, load the value of `ConfigurationNode#getString` as we parse currently.

MC-MrBirdy commented:
Changed to code based on your suggestion to use the `ConfigurationNode#childrenList`.
This results in a config as suggested:
```yml
    loot:
      - type: item
        material: diamond_axe
        weight: 3
        amount: 1
        nbt:
          damage_multiplier: 0.5
        enchantments:
        - name: efficiency
          level: '0-5'
          chance: 0.5
        - 'unbreaking 0-3'
```

I've added a new record type for `LootEnchantLevel` used by a function `parseLevel` to deal with the `levelString`.
This to keep `parseSingleEnchantList` hopefully more readable.

Archy-X reviewed (approved):
Thanks for implementing the request config format. This looks good to me!

Archy-X commented:
A reminder to use a branch in your fork and not your master branch, since when I squash and commit it will overwrite the history. I saw you used a branch in your latest PR but just letting you know in case.
